src/main/java/org/jitsi/jigasi/xmpp/CallControlMucActivator.java

```diff
@@ -0,0 +1,422 @@
+    {
+        osgiContext = bundleContext;
+
+        ConfigurationService config = getConfigurationService();
+
+        if (!config.getBoolean(BREWERY_ENABLED_PROP, false))
+        {
+            logger.warn("MUC call control disabled.");
+            return;
+        }
+
+        osgiContext.addServiceListener(this);
+
+        Collection<ServiceReference<ProtocolProviderService>> refs
+            = ServiceUtils.getServiceReferences(
+            osgiContext,
+            ProtocolProviderService.class);
+
+        for (ServiceReference<ProtocolProviderService> ref : refs)
+        {
+            initializeNewProvider(osgiContext.getService(ref));
+        }
+
+        SipGateway gateway = ServiceUtils.getService(
+            bundleContext, SipGateway.class);
+        if (gateway != null)
+        {
+            gateway.addSipGatewayListener(this);
+
+            if (this.callControl == null)
+                this.callControl = new CallControl(gateway, config);
+        }
+    }
+
+    /**
+     * Stops and unregister everything - providers, listeners.
+     * @param bundleContext
+     * @throws Exception
+     */
+    @Override
+    public void stop(BundleContext bundleContext)
+        throws Exception
+    {
+        osgiContext.removeServiceListener(this);
+
+        Collection<ServiceReference<ProtocolProviderService>> refs
+            = ServiceUtils.getServiceReferences(
+            osgiContext,
+            ProtocolProviderService.class);
+
+        for (ServiceReference<ProtocolProviderService> ref : refs)
+        {
+            ProtocolProviderService pps = osgiContext.getService(ref);
+            if (ProtocolNames.JABBER.equals(pps.getProtocolName()))
+            {
+                try
+                {
+                    pps.unregister();
+                }
+                catch(OperationFailedException e)
+                {
+                    logger.error("Cannot unregister xmpp provider", e);
+                }
+            }
+        }
+    }
+
+    /**
+     * Returns <tt>ConfigurationService</tt> instance.
+     * @return <tt>ConfigurationService</tt> instance.
+     */
+    public static ConfigurationService getConfigurationService()
+    {
+        return ServiceUtils.getService(
+            osgiContext, ConfigurationService.class);
+    }
+
+    @Override
+    public void serviceChanged(ServiceEvent serviceEvent)
+    {
+        if (serviceEvent.getType() != ServiceEvent.REGISTERED)
+            return;
+
+        ServiceReference ref = serviceEvent.getServiceReference();
+
+        Object service = osgiContext.getService(ref);
+
+        if (service instanceof ProtocolProviderService)
+        {
+            initializeNewProvider((ProtocolProviderService) service);
+        }
+        else if (service instanceof SipGateway)
+        {
+            SipGateway gateway = (SipGateway) service;
+            gateway.addSipGatewayListener(this);
+
+            if (this.callControl == null)
```

Comment:
if this.callControl !== null it would be good to check for reference equality here and print na error to detect multiple SipGateway instances. I suppose we expect to have only one instance at a time ?